### PR TITLE
Add a Global Styles endpoint and use it in the site editor

### PR DESCRIPTION
**Pull request by youknowriad**
This PR explores the global styles endpoint explored in https://github.com/WordPress/gutenberg/discussions/35141

At the moment, it only solves the "user global styles" use-cases, so there's no big difference with what we have on trunk aside of the fact that json parsing and serializing is moved to the server.

After that I'd like to explore:

 - ~getting rid of `__experimentalGlobalStylesUserEntityId`~ Done in 369183e
 - getting rid of `__experimentalGlobalStylesBaseConfig` 

The latter here might a bit complex because it will make the source for the base config is not a CPT but the theme file directly, so it's unclear to me at the moment whether it should be dedicated endpoints `global-styles/theme/{some_theme}` or whether we should try to merge things using the `id` like we did for templates (which feels too heavy for global styles)

### CODE REVIEW

File: packages/edit-site/src/components/global-styles/global-styles-provider.js
Comment (youknowriad) [inline packages/edit-site/src/components/global-styles/global-styles-provider.js:79]: I'm not sure how to handle this thing, I thought it was not something the client should be concerned about but I don't understand it enough to do the right things. cc @jorgefilipecosta

**Comment by youknowriad** [comment]
I've updated the PR and removed the `__experimentalGlobalStylesUserEntityId` flag. The only remaining thing for me is `__experimentalGlobalStylesBaseConfig` and it's not clear to me yet how to solve that: should it be part of the themes endpoint, a sub endpoint `/themes/{name}/global-styles` or should it be its own endpoint or merged with the global-styles endpoint with the smart "id" logic (like templates). I'd love your thoughts on this one @TimothyBJacobs @oandregal I'm thinking the first one might be the simplest for me (aka `/themes/{name}/global-styles`)

**Comment by oandregal** [comment]
Hey, took some time to read through the discussion and take stock of the data used by consumers:

global-styles-settings-data

They take the settings & styles, either standalone such as `user` or merged such as `core+theme+user` and do things with them. When it comes to generating the stylesheet from the given set of styles&settings, some of the consumers filter the data taking into account origins (core, theme, user) and types of styles (CSS variables, block classes, preset classes).

Additionally, note that some settings (presets) contain data for the three origins (i.e.: the data in `settings.__experimentalFeatures.color.palette` has `core`, `theme`, and `user` sub-keys).

We've also talked about "global styles variations" which I presume is just a list of different `core+theme` data.

**Comment by oandregal** [comment]
In the mental model I have for this, the endpoint would be something like:

`/wp/v2/global-styles/user/{ID}` => user data for the active theme
`/wp/v2/global-styles/base` => core+theme data for the active theme
`/wp/v2/global-styles/merged` => core+theme+user data for the active theme

This pattern allows us to grow later, should we need it:

`/wp/v2/global-styles/list` => a list of global styles variations that offer core+theme data for the active theme
`/wp/v2/global-styles/theme` => theme data for the active theme

In terms of naming: is there an alternative namespace to `global-styles` that communicates that data is both settings & styles?

**Comment by youknowriad** [comment]
> /wp/v2/global-styles/user/{ID} => user data for the active theme

This one is currently already implemented in this PR as ` /wp/v2/global-styles/{ID}`

> /wp/v2/global-styles/base => core+theme data for the active theme
> /wp/v2/global-styles/merged => core+theme+user data for the active theme

It is not clear to me whether these should be part of the same endpoint or more something specific to the themes endpoint

`/wp/v2/themes/{theme}/global-styles`

> /wp/v2/global-styles/list

This one is already implemented in this PR (kind of, just missing query support) `/wp/v2/global-styles`

I'd like some input from @TimothyBJacobs to get things as close as possible to the REST API guidelines.

**Comment by spacedmonkey** [comment]
This endpoint should really extend  post controller and not be a custom endpoint. We can override anything we need to on that controller.

**Comment by youknowriad** [comment]
> This endpoint should really extend post controller and not be a custom endpoint

I thought about that, it only works if we limit it to the CPT (only server user modified settings and styles) which is still uncertain. https://github.com/WordPress/gutenberg/pull/35801#issuecomment-949427340 and https://github.com/WordPress/gutenberg/pull/35801#issuecomment-949467023

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [inline lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: It's generally good practice to avoid unbounded queries like this and instead fully support pagination. Though in this case, is it even possible to accumulate thousands of these entries?

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [inline lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: Nit, but I'd expand this to an `if () {} else {}`

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [inline lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: We should also be checking `$request['title']['raw']`. You can see the posts controller for a reference.

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [inline lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]:
```suggestion
				'rendered' => get_the_title( $post ),
```

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [inline lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php:315]: We should define the possible `properties` here for `raw` and `rendered`.

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [inline lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: I think `styles` and `settings` should probably not be part of the `embed` response.

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [inline lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]:
```suggestion
					'default'     => [],
```

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [inline lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]:
```suggestion
					'default'     => [],
```

File: lib/compat/wordpress-5.9/rest-active-global-styles.php
Comment (TimothyBJacobs) [inline lib/compat/wordpress-5.9/rest-active-global-styles.php]:
```suggestion
			rest_url( 'wp/v2/global-styles/' . $global_styles_query->posts[0]->ID )
```

File: lib/compat/wordpress-5.9/rest-active-global-styles.php
Comment (TimothyBJacobs) [inline lib/compat/wordpress-5.9/rest-active-global-styles.php]: Link relations need to either be from the list of [IANA registered relations](https://www.iana.org/assignments/link-relations/link-relations.xhtml) or namespaced.

```suggestion
			'https://api.w.org/global-styles',
```

**Review by TimothyBJacobs — DISMISSED** [review]
Overall this looks good to me. I've left some feedback inline.

I think in this case, considering how custom the global styles API is, I think it is fine to extend the base controller instead of using `WP_REST_Posts_Controller`.

> > /wp/v2/global-styles/base => core+theme data for the active theme
> > /wp/v2/global-styles/merged => core+theme+user data for the active theme
> It is not clear to me whether these should be part of the same endpoint or more something specific to the themes endpoint

Yeah, my preference would be for this to be linked to by the themes endpoint.

> > /wp/v2/global-styles/list
> This one is already implemented in this PR (kind of, just missing query support) /wp/v2/global-styles

Yeah, the list should just be the collection endpoint.

**Comment by spacedmonkey** [comment]
> I think it is fine to extend the base controller instead of using WP_REST_Posts_Controller

Disagree. Much of this logic can be removed by extending this controller. Not to mention of all the filters and hooks in the post controller that are very useful.

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (youknowriad) [reply lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: I don't expect a big number of entries here. I was actually thinking we could start without the "list" method entirely as it's not useful for us yet.

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (youknowriad) [reply lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: Why? My personal preference is ternaries for assignments?

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (youknowriad) [reply lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: What is the "embed" context?

**Comment by youknowriad** [comment]
> /wp/v2/global-styles/base => core+theme data for the active theme
> /wp/v2/global-styles/merged => core+theme+user data for the active theme

Since these are going to be a separate endpoint, I'll leave them to their own PR.

For the Post Controller vs non Post Controller. I personally don't have a big preference here and it should be something we can change with impact on the API itself.

File: packages/edit-site/src/components/global-styles/global-styles-provider.js
Comment (jorgefilipecosta) [reply packages/edit-site/src/components/global-styles/global-styles-provider.js:79]: Yes, I agree that the client should not care about this validation now that we have an endpoint. I think on the endpoint method prepare_item_for_response we should check if the flag isGlobalStylesUserThemeJSON is present if it is we proceed normally if it is not we return empty styles and settings.

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [reply lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: IMO once we get to multiple lines for the construct, it's too big for a ternary. I know we do use them more frequently in JS, particularly in assignment because braces impact the scope, but we don't have those constraints in PHP.

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [reply lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: It is what is used when the resource is embedded via a `WP_REST_Response::add_link`.

**Comment by TimothyBJacobs** [comment]
> > I think it is fine to extend the base controller instead of using WP_REST_Posts_Controller
> 
> Disagree. Much of this logic can be removed by extending this controller. Not to mention of all the filters and hooks in the post controller that are very useful.

I'm not sure this is true. We'd need to be disabling essentially the entirety of how the posts controller works besides the title handling.

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (spacedmonkey) [inline lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php:25]: This should not be hard coded. Let's get the post type object and get it from there.

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (spacedmonkey) [inline lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: How about the using the post_content_filtered field?

**Comment by spacedmonkey** [comment]
What about meta and term handling? 

What about all the filters and actions in the post type controller?

**Comment by TimothyBJacobs** [comment]
> What about meta and term handling?
> 
> What about all the filters and actions in the post type controller?

I don't think we want to be supporting those at this point. Eventually, if we want to we can switch to the posts controller as . But for now, I think we want to keep this as minimal as possible.

**Review by jorgefilipecosta — COMMENTED** [review]
On my tests, I'm noticing a strange bug. My host is yah.local and the URL being requested is:
http://yah.local/wp-json/http://yah.local/wp-json/wp/v2/global-styles/32501?_locale=user which does not exist and returns an error. The URL requested should be http://yah.local/wp-json/wp/v2/global-styles/32501?_locale=user.

File: packages/core-data/src/entities.js
Comment (jorgefilipecosta) [inline packages/core-data/src/entities.js:134]: Should we pass baseURLParams: { context: 'edit' }? and plural: 'globalStyles' and include a key?

File: packages/core-data/src/resolvers.js
Comment (jorgefilipecosta) [inline packages/core-data/src/resolvers.js]: I guess this is the origin of the bug I identified here https://github.com/WordPress/gutenberg/pull/35801/#pullrequestreview-789738365 globalStylesURL is a full URL including host and HTTP part, apiFetch receives a path instead of a full URL and sp adds https://host/wp-json to the path that already contains making the request url invalid.

**Comment by jorgefilipecosta** [comment]
The link is to wp:global-styles, would it make sense for the link to be to wp:user-global-styles in case later we have distinct endpoints for the user and theme styles?

File: packages/edit-site/src/components/global-styles/global-styles-provider.js
Comment (jorgefilipecosta) [reply packages/edit-site/src/components/global-styles/global-styles-provider.js:79]: Related to this issue when serializing the user post to the database we should also include isGlobalStylesUserThemeJSON true, so the structure is properly sanitized. Right now we are not including the flag, so the styles load on the site editor, but on the frontend, user styles are ignored.

**Comment by jorgefilipecosta** [comment]
The direction o this PR seems good, I guess the PR can be merged after the issues identified ( https://github.com/WordPress/gutenberg/pull/35801#discussion_r736810038, https://github.com/WordPress/gutenberg/pull/35801#discussion_r736797810 )  are addressed. Nice work @youknowriad et all 👍 

> I've updated the PR and removed the __experimentalGlobalStylesUserEntityId flag. The only remaining thing for me is __experimentalGlobalStylesBaseConfig and it's not clear to me yet how to solve that: should it be part of the themes endpoint, a sub endpoint /themes/{name}/global-styles or should it be its own endpoint or merged with the global-styles endpoint with the smart "id" logic (like templates). I'd love your thoughts on this one @TimothyBJacobs @oandregal I'm thinking the first one might be the simplest for me (aka /themes/{name}/global-styles)

Regarding this, I think we should not expose the base styles or config anymore. We should have separate ways to request the theme theme.json and the core theme.json. For example, the presets need to be accessed per origin and the merge result is not enough I think we should have very simple endpoints that just return core and theme. For the theme, if we have child themes the return should be the result of the merge with the parent themes.

Well the paths of the theme.json files are static so we could already request them by their path and we already have some kind of endpoint but that seems very hacky and in the case of child, themes the hacky solution does not provide the result of the merge.

Something like /themes/{name}/global-styles as @youknowriad proposed may work for themes. I wonder how one would request the core theme.json? Maybe for the core, it is ok to use the hacky solution given that core theme.json does not require any merge, etc?

**Comment by TimothyBJacobs** [comment]
> Something like /themes/{name}/global-styles as @youknowriad proposed may work for themes.

Agreed.

>  I wonder how one would request the core theme.json? Maybe for the core, it is ok to use the hacky solution given that core theme.json does not require any merge, etc?

Sorry, to clarify, which solution are you referring to here?

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (youknowriad) [reply lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php:25]: Doing so will make a CPT REST API endpoint. I can do that but at the time of writing it's not considered as such, it's considered as an endpoint that is independent of the CPT and potentially could return things outside the CPT in the future.

I guess the only "visible" impact in terms of API would be that instead of doing 

```
getEntityRecords( 'root', 'globalStyles' );
```

we'd end up doing 

```
getEntityRecords( 'postType', 'wp_global_styles' ); 
```

in the frontend to retrieve the same data. Not sure which is best. Any thoughts here @jorgefilipecosta

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (youknowriad) [reply lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: I've used the same thing as the templates controller here but I can update, I'm not really clear what's the difference and what should be used here.

File: packages/core-data/src/resolvers.js
Comment (youknowriad) [reply packages/core-data/src/resolvers.js]: I've used url instead of path, I think it solves but would appreciate testing as I don't reproduce myself.

**Comment by oandregal** [comment]
Hey, I wanted to take a look at the naming of these endpoints all together. Given this and the related PR https://github.com/WordPress/gutenberg/pull/35985 it looks like this is what we're aiming for:

1. user data: `wp/v2/global-styles/{ID}` 
2. base data: `wp/v2/themes/${THEME}/global-styles`
3. merged data: `?`

I find the URL patterns not very clear: the first one only offers user data but is named as if was the whole merged data (core+theme+user); the second one is named after the theme but it actually returns core+theme data. It's also not clear to me how we're going to extend them should we need it: how are we naming a future endpoint that returns only the theme data for example?

An alternative would be a model where we only have core, theme, user endpoints (e.g.: `wp/v2/global-styles/user`, `wp/v2/global-styles/theme`, `wp/v2/global-styles/core`) as Jorge suggested. That's appealing, although it requires all the consumers to implement the merging logic (including mobile). I have a slight preference for not requiring this to consumers.

From the alternatives mentioned here, I'd still go with something along the lines of [this model](https://github.com/WordPress/gutenberg/pull/35801#issuecomment-949427340).

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [reply lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php:25]: Yeah, I don't think we should be grabbing this from the CPT object. This isn't a CPT being edited, but a global styles resource that is being stored into a CPT.

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (TimothyBJacobs) [reply lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php]: The customizer changeset uses `post_content` to store JSON. I think we should continue to use `post_content` to be consistent.

**Comment by jorgefilipecosta** [comment]
Hi @youknowriad,
I added a commit with changes:

Removed the defaults for settings and styles otherwise isset( $request['styles'] ) does not make sense. The request always contains styles. This change allows the users of the endpoints to change just settings or styles independently. 


We have a complex issue, when unserializing in PHP a json object as an array with `{settings: {} }` it becomes `{settings: [] }` and then the client crashes everywhere. I made many changes related to this to avoid this issue namely replacing some usages of array() with new stdClass(), replacing some isset with ! empty, and using cleanEmptyObject on the client which is good anyway to avoid wasting the database with empty object keys anyway (otherwise when the user for example resets the color palette we would pass `{settings: color: {} }` and on the new load the editor would crash because `{settings: color: [] }` is invalid.

I added the required isGlobalStylesUserThemeJSON flag verifications.

I added a hardcoded version (`config['version'] = WP_Theme_JSON_Gutenberg::LATEST_SCHEMA;`) when saving so the content properly loads on the frontend. This is a temporary solution to get things going we need to handle versions properly.
We need to make decisions like the endpoint is dumb and just saves/passes the version received to the database as is and the client is responsible for properly migrating things. Or the endpoint always receives and passes the latest version, (this means it is not back-compatible with clients relying on a previous theme.json shape). But in this solution the client does not need to implement any migration logic, if an old theme.json shape is saved on the database, the endpoint can rely on the migration the server already has to render styles and pass the latest version to the client (independently of what is on the database the client would always receive the latest shape).


There is a remaining issue. If a global styles post is not already on the database the site editor crashes right away. Previously the call WP_Theme_JSON_Resolver_Gutenberg::get_user_custom_post_type_id(); would create a post if it did not exist. We need a solution to this issue before merging. I guess a possible solution would be something like we had before and if $global_styles_query does not contain a post we would create a post so we can return it. But I guess a better solution would be to create a post only when saving some custom user styles.

**Comment by jorgefilipecosta** [comment]
> > Something like /themes/{name}/global-styles as @youknowriad proposed may work for themes.
> 
> Agreed.
> 
> > I wonder how one would request the core theme.json? Maybe for the core, it is ok to use the hacky solution given that core theme.json does not require any merge, etc?
> 
> Sorry, to clarify, which solution are you referring to here?

For the hacky solution to retrieve the core theme.json would be to make an HTTP request directly to the static path where the core default theme.json is located.

**Comment by jorgefilipecosta** [comment]
> Hey, I wanted to take a look at the naming of these endpoints all together. Given this and the related PR #35985 it looks like this is what we're aiming for:
> 
> 1. user data: `wp/v2/global-styles/{ID}`
> 2. base data: `wp/v2/themes/${THEME}/global-styles`
> 3. merged data: `?`
> 
> I find the URL patterns not very clear: the first one only offers user data but is named as if was the whole merged data (core+theme+user); the second one is named after the theme but it actually returns core+theme data. It's also not clear to me how we're going to extend them should we need it: how are we naming a future endpoint that returns only the theme data for example?
> 
> An alternative would be a model where we only have core, theme, user endpoints (e.g.: `wp/v2/global-styles/user`, `wp/v2/global-styles/theme`, `wp/v2/global-styles/core`) as Jorge suggested. That's appealing, although it requires all the consumers to implement the merging logic (including mobile). I have a slight preference for not requiring this to consumers.
> 
> From the alternatives mentioned here, I'd still go with something along the lines of [this model](https://github.com/WordPress/gutenberg/pull/35801#issuecomment-949427340).

The issue I see with this model: `wp/v2/global-styles/user`, `wp/v2/global-styles/theme`, `wp/v2/global-styles/core`, is that it is dependent of the current theme. I may want to retrieve the user styles of another theme besides the active one, or the theme styles of a different theme.


That said I imagine a something like `wp/v2/global-styles/user/{postid}`, `wp/v2/global-styles/theme/{theme-slug}`, `wp/v2/global-styles/core/` would work very well.
And if we need merged `wp/v2/global-styles/merged/{theme-slug}/{postid}` may also do the trick. But I'm not sure we need an endpoint for the merged styles. Mobile will need to have the merge logic anyway so the change the users are doing can be live previewed. And mobile can reuse our JS merge engine.

**Comment by youknowriad** [comment]
>  I guess a possible solution would be something like we had before and if $global_styles_query does not contain a post we would create a post so we can return it.

Yes, let's start with this solution first and iterate on it later if needed.

**Review by jorgefilipecosta — APPROVED** [review]
The issues I identified are fixed and I think this PR can be merged and when needed we can iterate on it 👍

File: lib/compat/wordpress-5.9/rest-active-global-styles.php
Comment (jorgefilipecosta) [inline lib/compat/wordpress-5.9/rest-active-global-styles.php]: I guess for this case given that get_user_custom_post_type_id already returns an id, we can avoid the duplicate queries that we are doing after.

File: lib/compat/wordpress-5.9/rest-active-global-styles.php
Comment (spacedmonkey) [inline lib/compat/wordpress-5.9/rest-active-global-styles.php:21]: This should be 
`'post_status'    =>  'publish',`

File: lib/compat/wordpress-5.9/rest-active-global-styles.php
Comment (spacedmonkey) [inline lib/compat/wordpress-5.9/rest-active-global-styles.php:34]: We could have just request `'fields' => 'ids'` in the WP_Query and no need for this `$global_styles_query->posts[0]->ID`, you could just do this. 

```
$id  = !empty( $global_styles_query->posts ) ? array_shift( $global_styles_query->posts ) : null;
```

File: lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php
Comment (spacedmonkey) [inline lib/compat/wordpress-5.9/class-gutenberg-rest-global-styles-controller.php:18]: What is this property used for?

**Comment by spacedmonkey** [comment]
I am confused. The global style post type is registered to the REST API. 

https://github.com/WordPress/gutenberg/blob/9953bfc6065f395eaa7f8ee43bf3265b53362ef4/lib/class-wp-theme-json-resolver-gutenberg.php#L444-L445

Is this REST API used for anything? If not this should be disabled no? CC @TimothyBJacobs ?

**Comment by youknowriad** [comment]
Great suggestions @spacedmonkey I opened a follow-up PR here to address these #36071

**Pull request merged**
